### PR TITLE
Bug fix for space leak on too frequent calls to performGC

### PR DESCRIPTION
**Pull request by schnecki**
see https://gitlab.haskell.org/ghc/ghc/-/issues/18239

possibly related https://github.com/hasktorch/hasktorch/issues/452

see also https://github.com/hasktorch/hasktorch/pull/453

### CODE REVIEW

File: hasktorch/src/Torch/Optim.hs [hunk 1]
```diff
@@ -23,6 +25,15 @@ newtype OptimizerState option = OptimizerState option
 grad' :: Loss -> [Parameter] -> Gradients
 grad' t p = Gradients (grad t p)
 
+-- | Action that performs garbage collection with a minimum waiting time between calls to `performGC`. This prevents a space leak, that occurs when calling `performGC` too frequently.
+performGC' :: IO ()
+performGC' = unsafePerformIO $ do
```
Comment (junjihashimoto) [inline hasktorch/src/Torch/Optim.hs:30]: I think unsafePerformIO should not be added.
It may disappear these codes.

**Comment by junjihashimoto** [comment]
@schnecki Thank you for your contribution!  Do you have any benchmarks?

**Comment by schnecki** [comment]
## Reproduction, benchmarks, and a follow-up commit

I reproduced the underlying pathology (GHC [#18239](https://gitlab.haskell.org/ghc/ghc/-/issues/18239)) and benchmarked the fix on the real optimizer path. I also pushed a small follow-up commit adding a `{-# NOINLINE performGC' #-}` pragma — rationale at the bottom.

TL;DR: the fix is correct and helps. On the **real** `runStep'` path it removes ~10–13 % of wall time with no downside; in the **isolated** GHC scenario the `performGC`-every-step pattern is catastrophic (≈**196×** slower).

### 1. Isolated GHC mechanism (`base` only)

`runStep'` calls `performGC` on *every* step. When a large structure is live on the GHC heap, each forced major GC re-traverses the whole live heap → O(steps × liveHeap). This script keeps ~256 MB live across 3000 steps and toggles the GC policy:

| policy | wall | per-step | bytes copied during GC | productivity |
|---|---:|---:|---:|---:|
| never `performGC` (baseline) | 0.39 s | ~0 ms | 0.69 GB | 16 % |
| **`performGC` every step (`runStep'`)** | **267.2 s** | **89.1 ms** | **806 GB** | **0.1 %** |
| debounced (this PR) | 1.36 s | 0.45 ms | 3.4 GB | 4 % |

`every` is **~196× slower** than debounced and copies **~237×** more GC bytes; 806 GB ≈ 3000 × 256 MB — exactly the O(steps × liveHeap) blow-up. Debounce (`deb`) matches the no-GC baseline.

### 2. Real hasktorch path (train a large MLP via the optimizer)

An `examples` executable trains a 2-hidden-layer MLP (`hidden=1024`, 500 steps, CPU) through an exact copy of `runStep'` with only the GC call parameterised (`every` / `deb` = `Control.Debounce` as in this PR / `never`):

| policy | wall | Δ vs deb | GC time | max GHC residency |
|---|---:|---:|---:|---:|
| never | 168.8 s | −2.6 % | 0.012 s | 0.40 MB |
| debounced (this PR) | 173.4 s | — | 0.047 s | 0.26 MB |
| **`performGC` every step (`runStep'`)** | **191.0 s** | **+10.2 %** | 0.110 s | 0.11 MB |

Note the tiny GHC residency (0.1–0.4 MB): tensor payloads live in libtorch's C++ heap, not the GHC heap, so the O(n²) *copy* blow-up from §1 does not dominate here. What remains is the per-step cost of forcing a major GC (finalizer/allocator churn) 500× — that alone is the ~10–13 % penalty, which debouncing removes cleanly (`deb` ≈ `never`). The §1 case is the general pathology this PR guards against once live GHC-heap data grows.

### 3. Follow-up commit: `NOINLINE performGC'`

`performGC'` is a CAF built with `unsafePerformIO (mkDebounce …)`. Without a `NOINLINE` pragma GHC may inline/duplicate the CAF, producing several independent debouncers that each keep their own timer — which defeats the rate limiting and reintroduces the leak. Added the pragma so the single debounced action is shared across all call sites.

### Scripts (self-contained, to reproduce)

<details><summary><code>scripts/gc_leak_repro.hs</code> — §1, base only: <code>ghc -O2 -rtsopts gc_leak_repro.hs &amp;&amp; ./gc_leak_repro 256 3000 1 +RTS -s</code></summary>

```haskell
{-# LANGUAGE BangPatterns #-}
module Main where

import           Control.Monad      (foldM, when)
import           Data.Array         (Array, listArray, bounds, (!))
import           Data.IORef
import           System.Environment (getArgs)
import           System.Mem         (performGC)
import           Text.Printf        (printf)
import           Data.Time.Clock    (getCurrentTime, diffUTCTime)

-- Big long-lived payload standing in for params + optimizer state. Boxed Array
-- so each element is a heap object the GC must trace, O(1)-indexable so we can
-- keep it genuinely live across the loop.
mkLive :: Int -> Array Int Double
mkLive mb = listArray (0, n - 1) [ fromIntegral i | i <- [0 .. n - 1] ]
  where n = max 1 (mb * 1024 * 1024 `div` 16)

-- One step: tiny work touching the live heap, then (optionally) a forced GC —
-- the shape of runStep'. gcEvery: 0=never, 1=every step, N=debounced.
oneStep :: Array Int Double -> Int -> IORef Double -> Int -> IO Double
oneStep live gcEvery acc i = do
    let (_, hi) = bounds live
        !x = live ! (i `mod` (hi + 1))
    modifyIORef' acc (\ !s -> s + x + fromIntegral (i `mod` 7))
    when (gcEvery > 0 && i `mod` gcEvery == 0) performGC
    readIORef acc

main :: IO ()
main = do
    args <- getArgs
    let liveMB  = argOr 0 128  args
        steps   = argOr 1 2000 args
        gcEvery = argOr 2 1    args
    let !live = mkLive liveMB
        (_, hi) = bounds live
    acc <- newIORef (0 :: Double)
    printf "liveMB=%d  liveElems=%d  steps=%d  gcEvery=%d\n" liveMB (hi + 1) steps gcEvery
    t0 <- getCurrentTime
    _  <- foldM (\_ i -> oneStep live gcEvery acc i) 0 [1 .. steps]
    t1 <- getCurrentTime
    final <- readIORef acc
    let wall = realToFrac (diffUTCTime t1 t0) :: Double
    printf "acc=%.1f\nwall=%.3fs  perStep=%.3fms\n" final wall (1000 * wall / fromIntegral steps)
  where
    argOr i d as = case drop i as of { (x:_) -> maybe d id (readMaybe x); _ -> d }
    readMaybe s = case reads s of { [(v,"")] -> Just v; _ -> Nothing }
```
</details>

<details><summary><code>examples/gc-perf-bench/Main.hs</code> — §2, real optimizer path (needs <code>libtorch-ffi</code>, <code>auto-update</code>, <code>time</code>). Run each: <code>gc-perf-bench 1024 500 &lt;every|deb|never&gt; +RTS -s</code></summary>

```haskell
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE ScopedTypeVariables #-}
module Main where

import           Control.Debounce   (defaultDebounceSettings, leadingEdge,
                                     debounceAction, debounceEdge, debounceFreq, mkDebounce)
import           Control.Monad      (foldM, when)
import           Data.Time.Clock    (diffUTCTime, getCurrentTime)
import           GHC.Generics       (Generic)
import           System.Environment (getArgs)
import           System.Mem         (performGC)
import           Text.Printf        (printf)
import           Torch
import           Torch.Internal.GC  (mallocTrim)

data MLP = MLP { l1 :: Linear, l2 :: Linear, l3 :: Linear } deriving Generic
instance Parameterized MLP

mlpForward :: MLP -> Tensor -> Tensor
mlpForward MLP{..} = linear l3 . relu . linear l2 . relu . linear l1

mkMLP :: Int -> Int -> IO MLP
mkMLP feats hidden =
  MLP <$> sample (LinearSpec feats hidden) <*> sample (LinearSpec hidden hidden)
      <*> sample (LinearSpec hidden 1)

mkGcAction :: String -> IO (IO ())
mkGcAction "every" = pure performGC
mkGcAction "never" = pure (pure ())
mkGcAction "deb"   = mkDebounce defaultDebounceSettings
  { debounceAction = performGC, debounceEdge = leadingEdge, debounceFreq = 1000000 }
mkGcAction other   = error ("unknown gc policy: " ++ other)

main :: IO ()
main = do
  args <- getArgs
  let hidden = argOr 0 2048 args :: Int
      steps  = argOr 1 1500 args :: Int
      policy = case drop 2 args of (x:_) -> x; _ -> "every"
      feats  = 256 :: Int; batch = 128 :: Int
  gcAction <- mkGcAction policy
  model0   <- mkMLP feats hidden
  gen0     <- mkGenerator (Device CPU 0) 31415
  let opt0 = mkAdam 0 0.9 0.999 (flattenParameters model0)
  printf "policy=%s  hidden=%d  steps=%d  feats=%d  batch=%d\n" policy hidden steps feats batch
  t0 <- getCurrentTime
  (final, _, _) <- foldM
    (\(m, o, g) i -> do
        let (x, g')  = randn' [batch, feats] g
            (tgt, _) = randn' [batch] g'
            loss     = mseLoss tgt (squeezeAll (mlpForward m x))
        (m', o') <- runStepWith gcAction m o loss 1e-4
        when (i `mod` 250 == 0) $ printf "  step %d  loss=%s\n" i (show (asValue loss :: Float))
        pure (m', o', g'))
    (model0, opt0, gen0) [1 .. steps]
  t1 <- getCurrentTime
  let wall = realToFrac (diffUTCTime t1 t0) :: Double
      liveParams = length (flattenParameters final)
  printf "wall=%.3fs  perStep=%.3fms  (params kept live: %d)\n"
         wall (1000 * wall / fromIntegral steps) liveParams
  where
    argOr i d as = case drop i as of { (x:_) -> maybe d id (readMaybe x); _ -> d }
    readMaybe s = case reads s of { [(v,"")] -> Just v; _ -> Nothing }

-- Copy of runStep' with the GC call made a parameter, so all three policies
-- share byte-for-byte identical optimizer work.
runStepWith :: (Parameterized model, Optimizer optimizer)
  => IO () -> model -> optimizer -> Loss -> LearningRate -> IO (model, optimizer)
runStepWith gcAction paramState optState lossValue lr = do
  let flatParameters = flattenParameters paramState
      depParameters  = fmap toDependent flatParameters
      gradients      = grad' lossValue flatParameters
  gcAction
  mallocTrim 0
  let (flatParameters', optState') = step lr gradients depParameters optState
  newFlatParam <- mapM makeIndependent flatParameters'
  pure (replaceParameters paramState newFlatParam, optState')
```
</details>

Environment: GHC 9.6.6, LTS 22.43, libtorch CPU, `auto-update-0.1.6`.

File: hasktorch/src/Torch/Optim.hs [hunk 2]
```diff
@@ -23,6 +25,21 @@ newtype OptimizerState option = OptimizerState option
...
+-- The @NOINLINE@ pragma is required: without it GHC may inline\/duplicate this
+-- CAF, yielding several independent debouncers that each track their own timer,
+-- which would defeat the rate limiting and reintroduce the space leak.
+performGC' :: IO ()
+performGC' = unsafePerformIO $ do
+  mkDebounce defaultDebounceSettings
+    { debounceAction = performGC
+    , debounceEdge = leadingEdge
+    , debounceFreq = 1000000
+    }
```
[the remark this answers is not in the document]
Comment (collinarnett) [reply hasktorch/src/Torch/Optim.hs:44]: Man I hate it when the models define a constant as a function in Haskell. Infuriating to deal with.

[the remark this answers is not in the document]
Comment (junjihashimoto) [reply hasktorch/src/Torch/Optim.hs:44]: @schnecki If you are dealing with models large enough to require GC every time, then frequent GC might indeed be necessary; however, since small models rarely require it, using a hard-coded constant is certainly not ideal—how about making it configurable via an environment variable instead?

[the remark this answers is not in the document]
Comment (junjihashimoto) [reply hasktorch/src/Torch/Optim.hs:44]: In the case of MLP, if performing GC every time only impacts performance by around 10%, it might be safe enough to leave it as is.
